### ISSUE TITLE
Offset is not uint

### ISSUE THREAD

**Issue by mkarajohn** [Borewit/music-metadata-browser#15]
## Issue description 

I am working on a music track uploader. The user is supposed to upload a music file, and the uploader is supposed to display the file metadata, once the upload is successful.

I have two mp3 files that both have the same kind of metadata, tracks A and B.

Track A's metadata are read and displayed fine. 

Track B's displays no metadata. 

I investigated on this, and the cause of the issue is that at some point, after the track has been translated into an array buffer, a `checkOffset` is called with an `offset` argument that is `undefined`, and as a result it throws (file location is `node_modules/node-libs-browser/node_modules/buffer/index.js` line 1097).

The best I could do, was check the initial array buffer for any `undefined` data inside of it, but the `.some()` I run on it returned `false`. 

The stack trace is this:

```
checkOffset (index.js:1165)
readUInt32LE (index.js:1218)
postMap (ID3v24TagMapper.js:194)
mapGenericTag (GenericTagMapper.js:40)
mapTag (CombinedTagMapper.js:54)
toCommon (MetadataCollector.js:256)
addTag (MetadataCollector.js:115)
addTag (ID3v2Parser.js:409)
_callee4$ (ID3v2Parser.js:357)
tryCatch (runtime.js:63)
invoke (runtime.js:282)
prototype.(anonymous function) (runtime.js:116)
fulfilled (ID3v2Parser.js:13)
Promise.then (async)
step (ID3v2Parser.js:30)
(anonymous) (ID3v2Parser.js:33)
push../node_modules/music-metadata/lib/id3v2/ID3v2Parser.js.__awaiter (ID3v2Parser.js:10)
parseId3Data (ID3v2Parser.js:170)
_callee$ (ID3v2Parser.js:95)
tryCatch (runtime.js:63)
invoke (runtime.js:282)
prototype.(anonymous function) (runtime.js:116)
fulfilled (ID3v2Parser.js:13)
Promise.then (async)
step (ID3v2Parser.js:30)
(anonymous) (ID3v2Parser.js:33)
push../node_modules/music-metadata/lib/id3v2/ID3v2Parser.js.__awaiter (ID3v2Parser.js:10)
parse (ID3v2Parser.js:59)
_callee2$ (AbstractID3Parser.js:162)
tryCatch (runtime.js:63)
invoke (runtime.js:282)
prototype.(anonymous function) (runtime.js:116)
fulfilled (AbstractID3Parser.js:19)
Promise.then (async)
step (AbstractID3Parser.js:36)
(anonymous) (AbstractID3Parser.js:39)
push../node_modules/music-metadata/lib/id3v2/AbstractID3Parser.js.__awaiter (AbstractID3Parser.js:16)
tryReadId3v2Headers (AbstractID3Parser.js:141)
_callee$ (AbstractID3Parser.js:105)
tryCatch (runtime.js:63)
invoke (runtime.js:282)
prototype.(anonymous function) (runtime.js:116)
(anonymous) (AbstractID3Parser.js:39)
push../node_modules/music-metadata/lib/id3v2/AbstractID3Parser.js.__awaiter (AbstractID3Parser.js:16)
parseID3v2 (AbstractID3Parser.js:96)
parse (AbstractID3Parser.js:83)
_callee3$ (ParserFactory.js:287)
tryCatch (runtime.js:63)
invoke (runtime.js:282)
prototype.(anonymous function) (runtime.js:116)
fulfilled (ParserFactory.js:13)
Promise.then (async)
step (ParserFactory.js:30)
(anonymous) (ParserFactory.js:33)
push../node_modules/music-metadata/lib/ParserFactory.js.__awaiter (ParserFactory.js:10)
_parse (ParserFactory.js:272)
_callee$ (ParserFactory.js:138)
tryCatch (runtime.js:63)
invoke (runtime.js:282)
prototype.(anonymous function) (runtime.js:116)
(anonymous) (ParserFactory.js:33)
push../node_modules/music-metadata/lib/ParserFactory.js.__awaiter (ParserFactory.js:10)
parse (ParserFactory.js:92)
parseFromTokenizer (core.js:56)
parseBuffer (core.js:37)
(anonymous) (index.js:61)
Promise.then (async)
parseBlob (index.js:60)
```

Can you direct me to the right direction for solving this?

I can provide you with the track files as well if you ask. 

Thank you, and thank you for this library.

**Comment by Borewit** [comment]
Thanks for reaching out @mkarajohn. 

> Can you direct me to the right direction for solving this?

I suspect some unexpected value in the `AverageLevel` or `PeakValue` tag, [might me a bug somewhere here](https://github.com/Borewit/music-metadata/blob/327e8e50e69f5407a2e2eec0f2032965f4184fb3/src/id3v2/ID3v24TagMapper.ts#L166-L174)

> I can provide you with the track files as well if you ask.

I would like to have a look to the track. I aim to convert any issue into a unit test. Which is the best way to debug the issue, put the file into a unit test, reproduce the problem, and run in through a debugger.

Enable  [the sourcemap for src](https://github.com/Borewit/music-metadata/blob/327e8e50e69f5407a2e2eec0f2032965f4184fb3/src/tsconfig.json#L4) to step through originating TypeScript.

I can fix it at well, just let know.

**Comment by mkarajohn** [comment]
>I suspect some unexpected value in the AverageLevel or PeakValue tag, might me a bug somewhere here

Ok, so I put a breakpoint there for `tag.value.owner_identifier === 'PeakValue'`, and I got the following findings:

* It never stops for the non-problematic track
* It stops for the problematic track, and then it [calls `readUInt32LE()`](https://github.com/Borewit/music-metadata/blob/327e8e50e69f5407a2e2eec0f2032965f4184fb3/src/id3v2/ID3v24TagMapper.ts#L172) with no arguments, and as a result `checkOffset()`  gets called with no arguments and the whole thing throws. 

```
Buffer.prototype.readUInt32LE = function readUInt32LE(offset, noAssert) {
  if (!noAssert) checkOffset(offset, 4, this.length);
  return (this[offset] | this[offset + 1] << 8 | this[offset + 2] << 16) + this[offset + 3] * 0x1000000;
};
```

The sample tracks are [here](https://drive.google.com/drive/folders/1Z8lUx_8BEKrrv9dMk4eO7TyC5M1VCDQe?usp=sharing). The problematic one is Beautiful Emptiness.

As a side note, I would like to get involved with this project, and I'd like to start by reading the source. However I am missing the greater context of how music tracks files are structured internally, so my question is, would you happen to have any pointers for things that I should read and get myself familiar with? (with regards to both music files and the code base in general)

**Comment by Borewit** [comment]
I am not able to reproduce your issue. You can try for yourself: Drag & drop the _Beautiful Emptiness_ file [in audio-tag-analyzer](https://audio-tag-analyzer.netlify.com/).

**Comment by Borewit** [comment]
Thanks a lot for the coffee, very much appreciated!

> As a side note, I would like to get involved with this project

That is great, welcome on board!

>  would you happen to have any pointers for things that I should read and get myself familiar with?

Hmm, hard to summarize in a few lines. 

Music-metadata is a stream based parsers, supported by [then-read-stream](https://github.com/Borewit/then-read-stream) to read portions of the a [readable stream](https://nodejs.org/api/stream.html#stream_class_stream_readable), and [strtok3](https://github.com/Borewit/strtok3) to read _tokens_ (token is encoded common data type like an integer or string). Each audio format (container) has its own specialized parser. Parsers may call shared parsers for common embedded data types, like ID3 headers.  Within the parser code you will typically find URLs to the corresponding descriptions.

Most audio containers are build up chunks, have a length and payload, very simular the way network messages are build up. MP3 (MPEG) format is an ugly format, as  a starting point, better to study how FLAC or MPEG-4 audio files are structured.

**Comment by Borewit** [comment]
You can ping me in the [chat](https://discord.gg/KyBr6sb) if you have questions.

**Comment by mkarajohn** [comment]
>I am not able to reproduce your issue.

You are right. I'll have to investigate some more then. I'll update in case this issue either needs to be closed or I manage to pinpoint the exact conditions for what I am describing in the OP.

Thanks for all the pointers!

**Comment by mkarajohn** [comment]
Ok, so after I created a minimum file uploader with metadata display on codesandbox.io (to compare with the implementation of the project at work) and after a few hours more of debugging, I finally reached the end of the rabbit hole.

The issue was a wrong dependency. Here's a breakdown:

* `music-metadata-browser` is depending on `buffer ^5.2.1`.  
* `webpack` (which we use at the project at work) is depending on `node-libs-browser ^2.0.0`.
* `node-libs-browser` is depending on `buffer ^4.3.0`.

The actual versions installed are as followed:

* `buffer@5.2.1` from `music-metadata-browser`, which is hoisted in `node_modules/` instead of `node_modules/music-metadata-browser/node_modules/`
* `node-libs-browser@2.2.0` from `webpack`, which is hoisted in `node_modules/` instead of `node_modules/webpack/node_modules/`
* `buffer@4.9.1` from `node-libs-browser`, which is in `node_modules/node-libs-browser/node_modules/`

The issue now, is that for some reason, which I am not sure why, `music-metadata-browser` is using `buffer@4.9.1` from inside `node_modules/node-libs-browser/node_modules/buffer`

`buffer@4.9.1` is lacking [this conversion](https://github.com/feross/buffer/blob/master/index.js#L1160) which was added in [this commit](https://github.com/feross/buffer/commit/99491d2f16d6f20940dd86ed2cb6e8629181723e#diff-168726dbe96b3ce427e7fedce31bb0bcR1100) back in 2016. As a result, `checkOffset()` gets called with `undefined` as the 1st parameter, and it throws. 

The project is using yarn. I tried deleting my `node_modules/` and `yarn.lock` and doing a fresh `yarn install` but the results were the same.

**Comment by Borewit** [comment]
Good catch! You could give `npm install` a try.

What if you checkout node-libs-browser and upgrade the buffer dependency?

You can depend on a local (modified) module using yarn using [yarn link](https://yarnpkg.com/lang/en/docs/cli/link/).

**Comment by Borewit** [comment]
> I'll update in case this issue either needs to be closed

I am not getting very nervous from an open issue, but in despite of the state, you can always update.

As long as issues has the same subject, and do not try to aim for more then a "single" change.

**Comment by mkarajohn** [comment]
I tried npm install too. Same results. I even tried npm install
--legacy-bundling, to try and force it not to flatten the deps, but this
way it doesn't even get to build successfully. The fact that the issue
persists with either yarn or npm gets me even more confused. I'll try
opening an issue on yarn anyway.

I'll also try your proposal as well tomorrow, thanks for your help.

**Comment by Borewit** [comment]
> I should probably report this to yarn

I am not convinced it is yarn causing the issue. It does the installation of the depend modules correct after all. I would investigate the module bundler.

**Comment by mkarajohn** [comment]
Ok, there is something vey strange happening, and I would like your assistance to confirm my findings.

[Here](https://codesandbox.io/s/8ko40vxrml) is the sandbox I created. It is very simple; it's a `create-react-app` project, the only dependency is `music-metadata-browser`, and all it does is show the cover art of the track you uploaded (use the 2 files from [here](https://drive.google.com/drive/folders/1Z8lUx_8BEKrrv9dMk4eO7TyC5M1VCDQe?usp=sharing))

As you can hopefully see, this example, running on codesandbox.io, works fine.

Now, here is the more involved part, as you will have to create a new `create-react-app` project, locally, on your machine. ([instructions here](https://github.com/facebook/create-react-app#quick-overview) in case you don't remember, so that you don't need to search).

Then copy paste, from codesandbox to your local project, the contents of the following files:
* `src/index.js`
* `package.json`

And finally `yarn/npm install` and `yarn start`. You do not need to eject.

I have executed the above on 3 separate machines (2 running macOS and 1 running Ubuntu) and the `buffer` package is always gets pulled from the wrong place. Check the included screenshot and see the path of `checkOffset()`. It should be pulled from `node_modules/buffer` instead it gets pulled from `node_modules/node-libs-browser/node_modules/buffer`

Can you confirm that it does not work locally for you either?

Thank you for your time.

**Comment by mkarajohn** [comment]
So here's a theory:

`music-metadata-browser` has `buffer: ^5.2.0` as a dependency. _However_, apart from importing the `Buffer` implementation once, for using it in [a type definition](https://github.com/Borewit/music-metadata-browser/blob/master/src/index.ts#L108) and never actually utilizes it. 

`music-metadata` uses the native Node `Buffer` implementation, on various occasions (without ever importing `buffer` of course, because it expects to be run in a Node environment, so `Buffer` is readily available)

Now, when `music-metadata-browser` passes the TypeScript compilation, the `import { Buffer } from 'buffer';` statement gets dropped, because the imported `Buffer`'s only use is for a type definition. So the final code only imports these packages:

```
const initDebug = require("debug");
const mm = require("music-metadata/lib/core");
const toBuffer = require("typedarray-to-buffer");
const Browser2NodeStream_1 = require("./fetch/Browser2NodeStream");
const debug = initDebug('music-metadata-browser');
```

instead of 

```
import { Buffer } from 'buffer';
import * as initDebug from 'debug';
import * as mm from 'music-metadata/lib/core';
import * as Type from 'music-metadata/lib/type';
import * as toBuffer from 'typedarray-to-buffer';
import { Browser2NodeStream } from './fetch/Browser2NodeStream';
```

And I think this might be the issue here: 

**Neither `music-metadata-browser` nor `music-metadata` final code ever _explicitly_ imports the `buffer` package**, so when the code gets compiled to run on the browser, **webpack defaults to provide any usages of the `Buffer` class with its own version**, which is included in `node-libs-browser`, and its `buffer@4.9.1`

**Comment by Borewit** [comment]
Impressive analysis, it looks like you nailed the issue. 
I have hardly any experience with react, so I very much appreciated your simple and clear guidance.:heart:

I tried both the sandbox, and the stand-alone project. 
It does indeed work in the sandbox and it doesn't on my end.

music-metadata is written for node.js, and uses the node api modules. music-metadata-browser is a wrapper aiming to adapt to requirements to run client side in the browser.

music-metadata-browser has indeed no real dependency on buffer, it imports buffer on behalf of music-metadata.

Which is in a way wrong. Problem is, putting it in music-metadata is also wrong, because buffer is provided by Node.js. It is the browser world which is borrowing technologies from the node.js world, like node modules. So adaptations, like polymorphing, I want to put on the browser side, and not the other way around.

I am pretty sure it should be possible to tweak webpack to use the correct buffer version, problem is that your react friends choose to hide it. I had similar challenges with Angular.

A small guide with this quality, to setup a react application with music-metadata would very helpful.
Or an hello music with metadata application in a sandbox would be great.

**Comment by mkarajohn** [comment]
>music-metadata-browser has indeed no real dependency on buffer, it imports buffer on behalf of music-metadata.

But that's the issue, it is not actually importing it 😄After Typescript gets compiled to plain Javascript, and type definitions no longer matter, the `import { Buffer } from 'buffer';` gets removed by the compiler from the code that actually runs on the client, because this import is not actually relevant to the final code (it's only relevant for [the type definition](https://github.com/Borewit/music-metadata-browser/blob/master/src/index.ts#L108) of Buffer)

So, while `music-metadata` in actuality _depends_ on a `buffer` version at least greater than [this](https://github.com/feross/buffer/commit/99491d2f16d6f20940dd86ed2cb6e8629181723e#diff-168726dbe96b3ce427e7fedce31bb0bcR1100), **when running in the browser**, it is not actually making sure that the correct `buffer` version (which is defined as ^5.2.x in its package.json) gets imported in the browser bundle. As a result webpack defaults to shim it with its own included version of `buffer` (which does not work with `music-metadata`)

>I am pretty sure it should be possible to tweak webpack to use the correct buffer version, problem is that your react friends choose to hide it.

Yeah, that can be done, our project at work is not a `create-react-app` project, so we have full control over all the configuration files. Still, I think that this is an "issue" with `music-metadata-browser`. 

My line of thinking is that since `music-metadata-browser' is a package that acts as a layer that allows `music-metadata` to run in a browser environment, it is _its_ responsibility to ensure that `music-metadata` will find all the necessary shims of node api modules, in the browser environment, and it should not be the responsibility of the individual to specifically configure webpack for it. 

One "stupid"/quick solution I can think of, is that in `music-metadata-browser` instead of importing `Buffer` from `buffer` just to use the `Buffer` in a type definition, make a small use of it, like a `Buffer.alloc()` or something,  in actual, code. 

This way the compiler will probably get tricked into thinking that `Buffer` is in fact used by the final code, and will not drop the `buffer@5.2` import, and webpack will see that import and use that version instead of its own.

I'll think of it better tomorrow, with a clear mind. Cheers.

**Comment by Borewit** [comment]
> So, while music-metadata in actuality depends on a buffer version at least greater than this

music-metadata depends on node (to provide the buffer). I think the above is only true in the context of music-metadata-browser / running in the browser.

> My line of thinking is that since music-metadata-browser is a package that acts as a layer that allows music-metadata to run in a browser environment

Indeed

> it is its responsibility to ensure that music-metadata will find all the necessary shims of node api modules, in the browser environment, and it should not be the responsibility of the individual to specifically configure webpack for it.

Webpack kicks in at the point where the node modules are assembled together in a bundled piece of data, closing the gap between the module system they borrowed from the node world into something which can be executed and efficiently provided to the browser. Music-metadata-browser is not at the end of that chain. It provides a node module designed to be used in a browser context. Browser folks also want their JavaScript to small, so a 2 times a buffer package is going to be an issue for others. So if I would preprocess I may interfere with bundler used in the application. Not only code size is an issue, it can also break stuff. I started like that actually and came to the conclusion that was not the solution.

I am open for improvements. I agree, as it is done now, it is not ideal neither. 

> My line of thinking is that since music-metadata-browser is a package that acts as a layer that allows music-metadata to run in a browser environment, it is its responsibility to ensure that music-metadata will find all the necessary shims of node api modules, in the browser environment, and it should not be the responsibility of the individual to specifically configure webpack for it.

If that does the trick, no problem. But it is not music-metadata-browser using buffer, issues may still pop in other modules designed for node.

It is the webpack / shim / polyfill which does the adaptation from node to browser. I would think that is the most logical place to address is. But think it is outside the code scope of this module, at the other end, I would like to make the use of this module easier.

**Comment by Borewit** [comment]
It looks like we addressed this issue in more constructive way in #28. Implemented a workaround for compatibility with node v4 in version [v1.2.2](https://github.com/Borewit/music-metadata-browser/releases/tag/v1.2.2)
